File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/oidc/TokenProvider.java
/*
 * Copyright © 2018 Knative Authors (knative-dev@googlegroups.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.knative.eventing.kafka.broker.core.oidc;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import dev.knative.eventing.kafka.broker.core.NamespacedName;
import io.fabric8.kubernetes.api.model.authentication.TokenRequest;
import io.fabric8.kubernetes.api.model.authentication.TokenRequestBuilder;
import io.fabric8.kubernetes.client.Config;
import io.fabric8.kubernetes.client.ConfigBuilder;
import io.fabric8.kubernetes.client.KubernetesClient;
import io.fabric8.kubernetes.client.KubernetesClientBuilder;
import java.util.concurrent.TimeUnit;

public class TokenProvider {

    private static final long TOKEN_EXPIRATION_SECONDS = 3600L; // 1 hour
    private static final long EXPIRATION_BUFFER_TIME_SECONDS = 300L; // 5 minutes

    private static final long CACHE_MAXIMUM_SIZE = 1000L; //  Cache up to 1000 tokens
    private static final long CACHE_EXPIRATION_TIME_SECONDS =
            TOKEN_EXPIRATION_SECONDS - EXPIRATION_BUFFER_TIME_SECONDS; // Cache tokens for 55 minutes

    private final KubernetesClient kubernetesClient;
    private final Cache<String, String> tokenCache;

    public TokenProvider() {
        Config clientConfig = new ConfigBuilder().build();
        kubernetesClient =
                new KubernetesClientBuilder().withConfig(clientConfig).build();

        this.tokenCache = CacheBuilder.newBuilder()
                .expireAfterWrite(CACHE_EXPIRATION_TIME_SECONDS, TimeUnit.SECONDS)
                .maximumSize(CACHE_MAXIMUM_SIZE)
                .build();
    }

    public String getToken(NamespacedName serviceAccount, String audience) {
        String cacheKey = generateCacheKey(serviceAccount, audience);
        String token = tokenCache.getIfPresent(cacheKey);

        if (token == null) {
            token = requestToken(serviceAccount, audience);
            if (token != null) {
                tokenCache.put(cacheKey, token);
            }
        }

        return token;
    }

    private String requestToken(NamespacedName serviceAccount, String audience) {
        TokenRequest tokenRequest = new TokenRequestBuilder()
                .withNewSpec()
                .withAudiences(audience)
                .withExpirationSeconds(TOKEN_EXPIRATION_SECONDS)
                .endSpec()
                .build();

        tokenRequest = kubernetesClient
                .serviceAccounts()
                .inNamespace(serviceAccount.namespace())
                .withName(serviceAccount.name())
                .tokenRequest(tokenRequest);

        if (tokenRequest != null && tokenRequest.getStatus() != null) {
            return tokenRequest.getStatus().getToken();
        } else {
            return null;
        }
    }

    private String generateCacheKey(NamespacedName serviceAccount, String audience) {
        return serviceAccount.namespace() + "/" + serviceAccount.name() + "/" + audience;
    }
}

File: control-plane/pkg/security/oauth/token_provider.go
/*
 * Copyright 2025 The Knative Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package oauth

import (
	"context"
	"fmt"

	"github.com/IBM/sarama"
)

// tokenIssuer defines the interface for generating access tokens
type tokenIssuer interface {
	issueToken(ctx context.Context) (string, error)
}

// TokenProvider provides common functionality for OAuth token providers
type TokenProvider struct {
	tokenIssuer tokenIssuer
}

// Token implements the sarama.AccessTokenProvider interface
func (b *TokenProvider) Token() (*sarama.AccessToken, error) {
	token, err := b.tokenIssuer.issueToken(context.TODO())
	if err != nil {
		return nil, err
	}
	return &sarama.AccessToken{Token: token}, nil
}

func NewTokenProvider(data map[string][]byte) (*TokenProvider, error) {
	tokenProvider, ok := data[saslTokenProviderKey]
	if !ok || len(tokenProvider) == 0 {
		return nil, fmt.Errorf("OAUTHBEARER token provider required (key: %s)", saslTokenProviderKey)
	}
	tokenProviderStr := string(tokenProvider)
	var tokenIssuer tokenIssuer
	var err error
	switch tokenProviderStr {
	case mskAccessTokenProvider:
		tokenIssuer, err = newMSKAccessTokenIssuer(data)
	case mskRoleAccessTokenProvider:
		tokenIssuer, err = newMSKRoleAccessTokenIssuer(data)
	default:
		return nil, fmt.Errorf("unsupported OAUTHBEARER token provider (key: %s), supported: %s, %s", saslTokenProviderKey, mskAccessTokenProvider, mskRoleAccessTokenProvider)
	}
	if err != nil {
		return nil, err
	}
	return &TokenProvider{
		tokenIssuer: tokenIssuer,
	}, nil
}

// errorTokenProvider is a sarama.AccessTokenProvider that always fails with a
// descriptive error. It is used when an OAUTHBEARER secret has no tokenProvider
// (the sasl.jaas.config / sasl.login.callback.handler.class keys are consumed by
// the Java data plane only). Setting this instead of leaving TokenProvider nil
// prevents a nil-pointer panic in sarama's OAUTHBEARER auth path when the Go
// control plane attempts an admin connection, turning it into a clear reconciler
// error instead.
type errorTokenProvider struct{ msg string }

func (e *errorTokenProvider) Token() (*sarama.AccessToken, error) {
	return nil, fmt.Errorf("%s", e.msg)
}

// UnsupportedTokenProvider returns an AccessTokenProvider that always fails with
// the given message. See errorTokenProvider for rationale.
func UnsupportedTokenProvider(msg string) sarama.AccessTokenProvider {
	return &errorTokenProvider{msg: msg}
}
